### ISSUE TITLE
Ensemble  Learning

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9298]
Ensemble  Learning<br>
<br>
https://ift.tt/aALQV6o<br>
<br>
贪钱算法还我头发<br>
<br>

<div><div><div><div><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br><span>9</span><br><span>10</span><br><span>11</span><br><span>12</span><br><span>13</span><br><span>14</span><br><span>15</span><br><span>16</span><br><span>17</span><br><span>18</span><br><span>19</span><br><span>20</span><br><span>21</span><br><span>22</span><br><span>23</span><br><span>24</span><br><span>25</span><br><span>26</span><br><span>27</span><br><span>28</span><br><span>29</span><br><span>30</span><br><span>31</span><br><span>32</span><br><span>33</span><br><span>34</span><br><span>35</span><br><span>36</span><br><span>37</span><br><span>38</span><br><span>39</span><br><span>40</span><br><span>41</span><br><span>42</span><br><span>43</span><br><span>44</span><br><span>45</span><br><span>46</span><br><span>47</span><br><span>48</span><br><span>49</span><br><span>50</span><br><span>51</span><br><span>52</span><br><span>53</span><br><span>54</span><br><span>55</span><br><span>56</span><br><span>57</span><br><span>58</span><br><span>59</span><br><span>60</span><br><span>61</span><br><span>62</span><br><span>63</span><br><span>64</span><br><span>65</span><br><span>66</span><br><span>67</span><br><span>68</span><br><span>69</span><br><span>70</span><br><span>71</span><br><span>72</span><br><span>73</span><br><span>74</span><br><span>75</span><br><span>76</span><br><span>77</span><br><span>78</span><br><span>79</span><br><span>80</span><br><span>81</span><br><span>82</span><br><span>83</span><br><span>84</span><br><span>85</span><br><span>86</span><br><span>87</span><br><span>88</span><br><span>89</span><br><span>90</span><br><span>91</span><br><span>92</span><br><span>93</span><br><span>94</span><br><span>95</span><br><span>96</span><br><span>97</span><br><span>98</span><br><span>99</span><br><span>100</span><br><span>101</span><br><span>102</span><br><span>103</span><br><span>104</span><br><span>105</span><br><span>106</span><br><span>107</span><br><span>108</span><br><span>109</span><br><span>110</span><br><span>111</span><br><span>112</span><br><span>113</span><br><span>114</span><br><span>115</span><br><span>116</span><br><span>117</span><br><span>118</span><br><span>119</span><br><span>120</span><br><span>121</span><br><span>122</span><br><span>123</span><br><span>124</span><br><span>125</span><br><span>126</span><br><span>127</span><br><span>128</span><br><span>129</span><br><span>130</span><br><span>131</span><br><span>132</span><br><span>133</span><br><span>134</span><br><span>135</span><br><span>136</span><br><span>137</span><br><span>138</span><br><span>139</span><br><span>140</span><br><span>141</span><br><span>142</span><br><span>143</span><br><span>144</span><br><span>145</span><br><span>146</span><br><span>147</span><br><span>148</span><br><span>149</span><br><span>150</span><br><span>151</span><br><span>152</span><br><span>153</span><br><span>154</span><br><span>155</span><br><span>156</span><br><span>157</span><br><span>158</span><br><span>159</span><br><span>160</span><br><span>161</span><br><span>162</span><br><span>163</span><br><span>164</span><br><span>165</span><br><span>166</span><br><span>167</span><br></pre></td>
<td><pre><span><span>import</span> imageio</span><br><span><span>import</span> numpy <span>as</span> np</span><br><span><span>import</span> matplotlib.pyplot <span>as</span> plt</span><br><span><span>from</span> matplotlib.colors <span>import</span> ListedColormap</span><br><span><span>from</span> sklearn.datasets <span>import</span> make_moons, make_circles</span><br><span><span>from</span> sklearn.tree <span>import</span> DecisionTreeClassifier</span><br><span><span>from</span> sklearn.metrics <span>import</span> f1_score</span><br><span><span>import</span> plotly.graph_objs <span>as</span> go</span><br><span><span>from</span> plotly.offline <span>import</span> download_plotlyjs, init_notebook_mode, plot, iplot</span><br><span></span><br><span></span><br><span><span><span>class</span> <span>Adaboost_Demonstration</span>:</span></span><br><span> <span><span>def</span> <span>__init__</span>(<span>self, X, y, learning_rate=<span>1.</span></span>):</span></span><br><span> <span>"""</span></span><br><span><span> 输入的X为N*2矩阵, y为一维向量, y的值只能取1或-1</span></span><br><span><span> :param X: 数据点</span></span><br><span><span> :param y: 数据点标记</span></span><br><span><span> """</span></span><br><span> self.X = X</span><br><span> self.y = y</span><br><span> </span><br><span> self.learning_rate = learning_rate</span><br><span> </span><br><span> self.num_samples = <span>len</span>(self.X)</span><br><span> </span><br><span> self.sample_weight = np.full(self.num_samples, <span>1</span> / self.num_samples)</span><br><span> </span><br><span> self.classifiers = []</span><br><span> </span><br><span> self.errors_list = []</span><br><span> </span><br><span> self.alphas = []</span><br><span></span><br><span> <span><span>def</span> <span>predict</span>(<span>self, data=<span>None</span>, labels=<span>None</span>, reduction=<span>"sign"</span></span>):</span></span><br><span> <span>"""</span></span><br><span><span> 预测数据点的分类</span></span><br><span><span> :param reduction: "sign"对弱分类的线性加权组合取符号, "mean"取平均</span></span><br><span><span> """</span></span><br><span> <span>if</span> data <span>is</span> <span>None</span>:</span><br><span> data = self.X</span><br><span> labels = self.y</span><br><span> </span><br><span> predictions = np.zeros([<span>len</span>(data)]).astype(<span>"float"</span>)</span><br><span> <span>for</span> classifier, alpha <span>in</span> <span>zip</span>(self.classifiers, self.alphas):</span><br><span> predictions += alpha * classifier.predict(data)</span><br><span> </span><br><span> <span>if</span> reduction == <span>"sign"</span>:</span><br><span> predictions = np.sign(predictions)</span><br><span> </span><br><span> <span>elif</span> reduction == <span>"mean"</span>:</span><br><span> predictions /= <span>len</span>(self.classifiers)</span><br><span> </span><br><span> <span>if</span> labels <span>is</span> <span>not</span> <span>None</span> <span>and</span> reduction == <span>"sign"</span>:</span><br><span> f1 = f1_score(predictions, labels)</span><br><span> <span>return</span> predictions, f1</span><br><span> <span>else</span>:</span><br><span> <span>return</span> predictions</span><br><span></span><br><span> <span><span>def</span> <span>contour_plot</span>(<span>self, data=<span>None</span>, labels=<span>None</span>, interval=<span>0.2</span>, title=<span>"adaboost"</span>, mode=<span>"3d"</span>, epoch=<span>None</span></span>):</span></span><br><span> <span>"""</span></span><br><span><span> 等高线图可视化</span></span><br><span><span> :param interval: 等高线图网格的间隔</span></span><br><span><span> :param title: 等高线图的标题</span></span><br><span><span> :param mode: 可选3D或2D可视化</span></span><br><span><span> :param epoch: 迭代次数</span></span><br><span><span> """</span></span><br><span> <span>if</span> data <span>is</span> <span>None</span>:</span><br><span> data = self.X</span><br><span> labels = self.y</span><br><span> <span>if</span> labels <span>is</span> <span>None</span>:</span><br><span> labels = np.ones([<span>len</span>(data)])</span><br><span> </span><br><span> x_min, x_max = data[:, <span>0</span>].<span>min</span>() - <span>.5</span>, data[:, <span>0</span>].<span>max</span>() + <span>.5</span></span><br><span> y_min, y_max = data[:, <span>1</span>].<span>min</span>() - <span>.5</span>, data[:, <span>1</span>].<span>max</span>() + <span>.5</span></span><br><span> xx, yy = np.meshgrid(np.arange(x_min, x_max, interval), np.arange(y_min, y_max, interval))</span><br><span> </span><br><span> X_grid = np.concatenate([np.expand_dims(np.ravel(xx), axis=-<span>1</span>),</span><br><span> np.expand_dims(np.ravel(yy), axis=-<span>1</span>)], axis=-<span>1</span>)</span><br><span> </span><br><span> </span><br><span> Z_grid = self.predict(data=X_grid, reduction=<span>"mean"</span>)</span><br><span> Z_grid = Z_grid.reshape(xx.shape)</span><br><span> </span><br><span> <span>if</span> mode == <span>"3d"</span>:</span><br><span> </span><br><span> scatter = go.Scatter3d(x=data[:, <span>0</span>], y=data[:, <span>1</span>], z=self.predict(data=data, reduction=<span>"mean"</span>),</span><br><span> mode=<span>'markers'</span>,</span><br><span> marker=<span>dict</span>(color=labels, size=<span>5</span>, symbol=<span>'circle'</span>,</span><br><span> line=<span>dict</span>(color=<span>'rgb(204, 204, 204)'</span>, width=<span>1</span>),</span><br><span> opacity=<span>0.9</span>))</span><br><span> </span><br><span> surface = go.Surface(x=xx, y=yy, z=Z_grid, opacity=<span>0.9</span>)</span><br><span> plot_data = [scatter, surface]</span><br><span> layout = go.Layout(title=title)</span><br><span> </span><br><span> camera = <span>dict</span>(up=<span>dict</span>(x=<span>0</span>, y=<span>0</span>, z=<span>1</span>),</span><br><span> center=<span>dict</span>(x=<span>0</span>, y=<span>0</span>, z=<span>0</span>),</span><br><span> eye=<span>dict</span>(x=<span>1</span>, y=<span>1</span>, z=<span>0.8</span>))</span><br><span> fig = go.Figure(data=plot_data, layout=layout)</span><br><span> fig[<span>'layout'</span>].update(scene=<span>dict</span>(camera=camera))</span><br><span> iplot(fig, image=<span>"png"</span>, filename=title)</span><br><span> <span>if</span> mode == <span>"2d"</span>:</span><br><span> </span><br><span> plt.contourf(xx, yy, Z_grid, cmap=plt.cm.RdBu, alpha=<span>.8</span>)</span><br><span> </span><br><span> plt.scatter(data[:, <span>0</span>], data[:, <span>1</span>], c=labels,</span><br><span> cmap=ListedColormap([<span>'#FF0000'</span>, <span>'#0000FF'</span>]), edgecolors=<span>'k'</span>)</span><br><span> plt.title(title)</span><br><span> </span><br><span> plt.savefig(<span>'imgs/'</span> + <span>'res'</span> + <span>str</span>(epoch))</span><br><span></span><br><span></span><br><span> <span><span>def</span> <span>__next__</span>(<span>self, reduction=<span>"mean"</span>, plot=<span>True</span>, plot_mode=<span>"2d"</span>, epoch=<span>None</span></span>):</span></span><br><span> </span><br><span> </span><br><span> </span><br><span> </span><br><span> classifier = DecisionTreeClassifier(max_depth=<span>1</span>)</span><br><span> </span><br><span> classifier.fit(self.X, self.y, sample_weight=self.sample_weight)</span><br><span> </span><br><span> predictions = classifier.predict(self.X)</span><br><span> </span><br><span> error_rate = np.mean(np.average((predictions != self.y), weights=self.sample_weight))</span><br><span> </span><br><span> alpha = self.learning_rate * (np.log((<span>1</span> - error_rate) / error_rate)) / <span>2</span></span><br><span> </span><br><span> self.sample_weight *= np.exp(-alpha * self.y * predictions)</span><br><span> </span><br><span> self.sample_weight /= np.<span>sum</span>(self.sample_weight)</span><br><span> </span><br><span> self.classifiers.append(classifier)</span><br><span> </span><br><span> self.alphas.append(alpha)</span><br><span> </span><br><span> _, f1 = self.predict()</span><br><span> </span><br><span> <span>if</span> plot:</span><br><span> <span>return</span> self.contour_plot(</span><br><span> title=<span>"adaboost step "</span> + <span>str</span>(<span>len</span>(self.classifiers)) + <span>" f1 score: {:.2f}"</span>.<span>format</span>(f1), mode=plot_mode, epoch=epoch)</span><br><span> <span>else</span>:</span><br><span> <span>return</span> f1</span><br><span></span><br><span></span><br><span><span><span>def</span> <span>convert_gif</span>(<span>epochs</span>):</span></span><br><span> outfile = <span>'gifs/adaboost.gif'</span></span><br><span> filenames = []</span><br><span> <span>for</span> i <span>in</span> <span>range</span>(epochs):</span><br><span> filename = <span>'imgs/res'</span> + <span>str</span>(i) + <span>'.png'</span></span><br><span> filenames.append(filename)</span><br><span> frames = []</span><br><span> <span>for</span> image_name <span>in</span> filenames:</span><br><span> img = imageio.imread(image_name)</span><br><span> frames.append(img)</span><br><span> imageio.mimsave(outfile, frames, <span>'GIF'</span>, duration=<span>'0.1'</span>)</span><br><span></span><br><span></span><br><span><span>if</span> __name__ == <span>'__main__'</span>:</span><br><span> </span><br><span> X, y = make_moons(n_samples=<span>300</span>, noise=<span>0.2</span>, random_state=<span>3</span>)</span><br><span> y[np.where(y == <span>0</span>)] = -<span>1</span></span><br><span> model = Adaboost_Demonstration(X, y)</span><br><span> epochs = <span>100</span></span><br><span> <span>for</span> i <span>in</span> <span>range</span>(epochs):</span><br><span> model.__next__(plot=<span>False</span>, epoch=i)</span><br><span> model.contour_plot(mode=<span>"2d"</span>)</span><br><span> </span><br></pre></td>
</tr></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via 一直进步 做喜欢的 <br>
<br>
May 2, 2023 at 08:31AM
